File: Queue/CircularQueue/CircularQueue.java
public class CircularQueue {
    private int[] data;
    private int front;
    private int rear;
    private int size;
    private int capacity;

    public CircularQueue(int capacity) {
        this.capacity = capacity;
        this.data = new int[capacity];
        this.front = 0;
        this.rear = -1;
        this.size = 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean enqueue(int value) {
        if (isFull()) {
            System.out.println("Queue Overflow! Cannot enqueue " + value);
            return false;
        }
        rear = (rear + 1) % capacity;
        data[rear] = value;
        size++;
        return true;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue Underflow! Cannot dequeue");
            return -1;
        }
        int value = data[front];
        front = (front + 1) % capacity;
        size--;
        return value;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return -1;
        }
        return data[front];
    }

    public void print() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return;
        }
        System.out.print("Queue (Front -> Rear): ");
        for (int i = 0; i < size; i++) {
            int idx = (front + i) % capacity;
            System.out.print(data[idx] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        CircularQueue q = new CircularQueue(4);
        q.enqueue(10);
        q.enqueue(20);
        q.enqueue(30);
        q.dequeue();
        q.enqueue(40);
        q.print(); // 20 30 40
    }
}

File: Queue/CircularQueue/CircularQueue.cpp
#include <iostream>

template <typename T, int Capacity>
class CircularQueue {
private:
    T data[Capacity];
    int front_idx;
    int rear_idx;
    int current_size;

public:
    CircularQueue() : front_idx(0), rear_idx(-1), current_size(0) {}

    bool isFull() const {
        return current_size == Capacity;
    }

    bool isEmpty() const {
        return current_size == 0;
    }

    bool enqueue(const T& value) {
        if (isFull()) {
            std::cout << "Queue Overflow! Cannot enqueue " << value << std::endl;
            return false;
        }
        rear_idx = (rear_idx + 1) % Capacity;
        data[rear_idx] = value;
        current_size++;
        return true;
    }

    T dequeue() {
        if (isEmpty()) {
            std::cout << "Queue Underflow! Cannot dequeue" << std::endl;
            return T();
        }
        T value = data[front_idx];
        front_idx = (front_idx + 1) % Capacity;
        current_size--;
        return value;
    }

    T peek() const {
        if (isEmpty()) {
            std::cout << "Queue is empty!" << std::endl;
            return T();
        }
        return data[front_idx];
    }

    void print() const {
        if (isEmpty()) {
            std::cout << "Queue is empty" << std::endl;
            return;
        }
        std::cout << "Queue (Front -> Rear): ";
        for (int i = 0; i < current_size; ++i) {
            int idx = (front_idx + i) % Capacity;
            std::cout << data[idx] << " ";
        }
        std::cout << std::endl;
    }
};

int main() {
    CircularQueue<int, 4> q;
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);
    q.dequeue();
    q.enqueue(40);
    q.print(); // 20 30 40
    return 0;
}
